File: Python/다이어리 프로그램.py
import getpass
from datetime import datetime, timedelta

class User:
    def __init__(self, name, username, password):
        self.name = name
        self.username = username
        self.password = password
        self.diary_entries = {}

class DiaryProgram:
    def __init__(self):
        self.users = {}
        self.current_user = None

    def run(self):
        while True:
            print("1. Login")
            print("2. Sign Up")
            choice = input("Select an option: ")

            if choice == '1':
                self.login()
                if self.current_user:
                    self.main_page()
            elif choice == '2':
                self.sign_up()

    def login(self):
        print("\n--- Login ---")
        username = input("Enter your username: ")
        password = getpass.getpass("Enter your password: ")

        if username in self.users and self.users[username].password == password:
            self.current_user = self.users[username]
            print("Login successful!")
        else:
            print("Invalid username or password.")

    def sign_up(self):
        print("\n--- Sign Up ---")
        name = input("Enter your name: ")
        username = input("Enter a username: ")

        if username.isdigit():
            print("Username cannot be all digits.")
            return

        if username in self.users:
            print("Username already exists.")
            return

        password = getpass.getpass("Enter a password: ")
        if not name or not username or not password:
            print("All fields are required.")
            return

        self.users[username] = User(name, username, password)
        print("Sign up successful!")

    def main_page(self):
        while True:
            print(f"\nWelcome, {self.current_user.name}'s Diary")
            today = datetime.now()
            self.display_calendar(today)

            print("3. My Diary")
            print("4. My Diary List")
            print("5. Logout")

            choice = input("Select an option: ")

            if choice == '3':
                self.diary_page(today)
            elif choice == '4':
                self.diary_list_page()
            elif choice == '5':
                self.current_user = None
                return

    def display_calendar(self, date):
        month = date.month
        year = date.year
        num_days = (date.replace(day=28) + timedelta(days=4)).day

        print(f"\n{date.strftime('%B')} {year}")
        print("  M  T  W  T  F  S  S")

        offset = date.weekday()
        for i in range(offset):
            print("   ", end=" ")

        for day in range(1, num_days + 1):
            if day == datetime.now().day and month == datetime.now().month:
                print(f"\033[91m{day:2}\033[00m ", end=" ")
            else:
                print(f"{day:2} ", end=" ")

            offset += 1
            if offset % 7 == 0:
                print()

    def diary_page(self, date):
        print(f"\n--- Diary Entry: {date.strftime('%B %d, %Y')} ---")
        entry = input("Write your diary entry (Press Enter to finish):\n")
        if entry.strip():
            self.current_user.diary_entries[date] = entry
            print("Entry saved successfully!")
        else:
            print("No entry to save.")

    def diary_list_page(self):
        print("\n--- Diary List ---")
        for date in sorted(self.current_user.diary_entries.keys()):
            print(date.strftime('%B %d, %Y'))

        selected_date = input("Select a date to view entry (MM/DD/YYYY), or type 'back' to return: ")
        if selected_date.lower() == 'back':
            return

        try:
            selected_date = datetime.strptime(selected_date, '%m/%d/%Y')
            if selected_date in self.current_user.diary_entries:
                print(f"\n--- Diary Entry: {selected_date.strftime('%B %d, %Y')} ---")
                print(self.current_user.diary_entries[selected_date])
            else:
                print("No entry found for selected date.")
        except ValueError:
            print("Invalid date format.")

if __name__ == "__main__":
    diary_program = DiaryProgram()
    diary_program.run()

File: Cpp/다이어리 프로그램.cpp
#include <iostream>
#include <vector>
#include <string>
#include <ctime>

#define MAX_ENTRIES 100

struct User {
    std::string name;
    std::string username;
    std::string password;
    std::vector<tm> diaryEntries;
    std::vector<std::string> entries;
};

void displayCalendar(const tm& date) {
    int month = date.tm_mon + 1;
    int year = date.tm_year + 1900;
    int numDays = 31; // You may need to calculate this based on the month and year

    std::cout << (month == 1 ? "January" : month == 2 ? "February" : month == 3 ? "March" : month == 4 ? "April" : month == 5 ? "May" : month == 6 ? "June" : month == 7 ? "July" : month == 8 ? "August" : month == 9 ? "September" : month == 10 ? "October" : month == 11 ? "November" : "December") << " " << year << std::endl;
    std::cout << "  M  T  W  T  F  S  S" << std::endl;

    // Implement the calendar display logic here
}

void diaryPage(User& user, const tm& date) {
    std::cout << "\n--- Diary Entry: " << date.tm_mon + 1 << "-" << date.tm_mday << "-" << date.tm_year + 1900 << " ---" << std::endl;
    std::cout << "Write your diary entry (Press Enter to finish):" << std::endl;
    std::string entry;
    std::getline(std::cin, entry);

    if (!entry.empty()) {
        user.entries.push_back(entry);
        user.diaryEntries.push_back(date);
        std::cout << "Entry saved successfully!" << std::endl;
    }
    else {
        std::cout << "No entry to save." << std::endl;
    }
}

void diaryListPage(const User& user) {
    std::cout << "\n--- Diary List ---" << std::endl;
    for (const auto& entry : user.diaryEntries) {
        std::cout << entry.tm_mon + 1 << "-" << entry.tm_mday << "-" << entry.tm_year + 1900 << std::endl;
    }
}

void mainPage(User& currentUser) {
    int choice;
    while (true) {
        std::cout << "\nWelcome, " << currentUser.name << "'s Diary" << std::endl;
        tm today = { 0 }; // Current date initialization
        displayCalendar(today);

        std::cout << "3. My Diary\n";
        std::cout << "4. My Diary List\n";
        std::cout << "5. Logout\n";
        std::cout << "Select an option: ";
        std::cin >> choice;
        std::cin.ignore(); // Consume the newline character left in the input buffer

        switch (choice) {
        case 3:
            diaryPage(currentUser, today);
            break;
        case 4:
            diaryListPage(currentUser);
            break;
        case 5:
            return; // Return to the main menu
        default:
            std::cout << "Invalid option." << std::endl;
        }
    }
}

void login(std::vector<User>& users, User*& currentUser) {
    std::cout << "\n--- Login ---" << std::endl;
    std::string username, password;
    std::cout << "Enter your username: ";
    std::getline(std::cin, username);
    std::cout << "Enter your password: ";
    std::getline(std::cin, password);

    for (auto& user : users) {
        if (user.username == username && user.password == password) {
            currentUser = &user;
            std::cout << "Login successful!" << std::endl;
            return;
        }
    }
    std::cout << "Invalid username or password." << std::endl;
}

void signUp(std::vector<User>& users) {
    std::cout << "\n--- Sign Up ---" << std::endl;
    User newUser;
    std::cout << "Enter your name: ";
    std::getline(std::cin, newUser.name);
    std::cout << "Enter a username: ";
    std::getline(std::cin, newUser.username);

    // Check if username already exists

    std::cout << "Enter a password: ";
    std::getline(std::cin, newUser.password);

    users.push_back(newUser);
}

int main() {
    std::vector<User> users; // Vector to store users
    User* currentUser = nullptr;
    int choice;

    while (true) {
        std::cout << "1. Login\n";
        std::cout << "2. Sign Up\n";
        std::cout << "3. Quit\n";
        std::cout << "Select an option: ";
        std::cin >> choice;
        std::cin.ignore(); // Consume the newline character left in the input buffer

        switch (choice) {
        case 1:
            login(users, currentUser);
            if (currentUser != nullptr) {
                mainPage(*currentUser);
            }
            break;
        case 2:
            signUp(users);
            break;
        case 3:
            std::cout << "Goodbye!" << std::endl;
            return 0; // 프로그램 종료
        default:
            std::cout << "Invalid option." << std::endl;
        }
    }

    return 0;
}
